File: 2015/Day9/src/day9_ai.py
from itertools import permutations

def parse_distances(file_path):
    # Initialize distance dictionary
    distances = {}
    cities = set()
    
    # Read from input.txt
    with open(file_path, 'r') as file:
        for line in file:
            parts = line.strip().split()
            if len(parts) != 5 or parts[1] != 'to' or parts[3] != '=':
                continue  # Skip malformed lines
            city1, city2, distance = parts[0], parts[2], int(parts[4])
            cities.add(city1)
            cities.add(city2)
            distances[(city1, city2)] = distance
            distances[(city2, city1)] = distance  # Undirected graph
    
    return distances, list(cities)

def tsp_shortest_longest(file_path):
    distances, cities = parse_distances(file_path)
    n = len(cities)
    
    # Get all possible permutations of cities
    min_distance = float('inf')
    max_distance = float('-inf')
    
    for perm in permutations(cities):
        total_distance = 0
        valid = True
        # Calculate total distance for the permutation
        for i in range(n-1):
            if (perm[i], perm[i+1]) not in distances:
                valid = False
                break
            total_distance += distances[(perm[i], perm[i+1])]
        
        if valid:
            min_distance = min(min_distance, total_distance)
            max_distance = max(max_distance, total_distance)
    
    return min_distance, max_distance

# Read from input.txt and compute results
shortest, longest = tsp_shortest_longest('input.txt')
print(f"Part 1 - Shortest route: {shortest}")
print(f"Part 2 - Longest route: {longest}")

File: 2015/Day9/src/day9_ai.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <map>
#include <set>
#include <string>
#include <algorithm>
#include <climits>

using namespace std;

// Parse input from file into a distance map and list of cities
pair<map<pair<string, string>, int>, vector<string>> parse_distances(const string& file_path) {
    map<pair<string, string>, int> distances;
    set<string> cities;
    ifstream file(file_path);
    string line;

    if (!file.is_open()) {
        cerr << "Error: Could not open input.txt" << endl;
        exit(1);
    }

    while (getline(file, line)) {
        string city1, city2, to;
        int distance;
        stringstream ls(line);
        if (ls >> city1 >> to >> city2 >> to >> distance) {
            cities.insert(city1);
            cities.insert(city2);
            distances[{city1, city2}] = distance;
            distances[{city2, city1}] = distance; // Undirected graph
        }
    }

    file.close();
    vector<string> city_list(cities.begin(), cities.end());
    return {distances, city_list};
}

// Compute shortest and longest TSP routes
pair<int, int> tsp_shortest_longest(const string& file_path) {
    auto [distances, cities] = parse_distances(file_path);
    int n = cities.size();
    vector<int> indices(n);
    for (int i = 0; i < n; ++i) indices[i] = i;

    int min_distance = INT_MAX;
    int max_distance = INT_MIN;

    do {
        int total_distance = 0;
        bool valid = true;
        for (int i = 0; i < n - 1; ++i) {
            string city1 = cities[indices[i]];
            string city2 = cities[indices[i + 1]];
            auto edge = distances.find({city1, city2});
            if (edge == distances.end()) {
                valid = false;
                break;
            }
            total_distance += edge->second;
        }
        if (valid) {
            min_distance = min(min_distance, total_distance);
            max_distance = max(max_distance, total_distance);
        }
    } while (next_permutation(indices.begin(), indices.end()));

    return {min_distance, max_distance};
}

int main() {
    auto [shortest, longest] = tsp_shortest_longest("input.txt");
    cout << "Part 1 - Shortest route: " << shortest << endl;
    cout << "Part 2 - Longest route: " << longest << endl;

    return 0;
}